src/contract.py

```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
+        if dtype not in [PutCallFwd.CALL, PutCallFwd.PUT]:
+            self._raise_incorrect_derivative_type()
+        super().__init__(und, dtype, longshort, strk, exp)
+        self._contract = ContractType.EUROPEANDIGITALOPTION
+
+    def convert_to_generic(self, contract, und, dtype, longshort, strk, exp) -> GenericContract:
+        return GenericContract(contract, und, dtype, longshort, strk, exp)
 
     def payoff(self, spot: float) -> float:
         if self._derivative_type == PutCallFwd.CALL:
-            return max(spot - self._strike, 0)
+            return self.ls * float(spot - self._strike > 0)
         elif self._derivative_type == PutCallFwd.PUT:
-            return max(self._strike - spot, 0)
+            return self.ls * float(self._strike - spot > 0)
         else:
             self._raise_incorrect_derivative_type()
 
 
-class EuropeanDigitalContract(VanillaOptionContract):
+class PathDependentContract(Contract):
+
+    def to_dict(self) -> dict[str, any]:
+        return super().to_dict()
+
+    @abstractmethod
     def convert_to_generic(self) -> GenericContract:
-        # todo: needs to be properly initialized after GenericContract is implemented
-        return GenericContract()
+        pass
 
+    @abstractmethod
     def payoff(self, spot: float) -> float:
+        pass
+
+    def get_timeline(self) -> list[float]:
+        return [((i+1)/self._num_mon)*self._expiry for i in range(self._num_mon)]
+
+
+class AsianContract(PathDependentContract):
+
+    def __init__(self, und: Stock, dtype: PutCallFwd, longshort: LongShort, strk: float, exp: float) -> None:
+        if dtype not in [PutCallFwd.CALL, PutCallFwd.PUT]:
+            self._raise_incorrect_derivative_type()
+        super().__init__(und, dtype, longshort, strk, exp)
+        self._contract = ContractType.ASIANOPTION
+
+    def convert_to_generic(self, contract, und, dtype, longshort, strk, exp) -> GenericContract:
+        return GenericContract(contract, und, dtype, longshort, strk, exp)
+
+    def payoff(self, prices_und: float) -> float:
+    # TO DO: prices_und to derive from the underlying process using the timeline
         if self._derivative_type == PutCallFwd.CALL:
-            return float(spot - self._strike > 0)
+            return self.ls * max(mean(prices_und) - self._strike, 0)
         elif self._derivative_type == PutCallFwd.PUT:
-            return float(self._strike - spot > 0)
+            return self.ls * max(self._strike - mean(prices_und), 0)
         else:
             self._raise_incorrect_derivative_type()
 
 
-class GenericContract:
-    # todo: needs to be implemented, defined temporarily as empty class due to annotations
-    pass
+class EuropeanBarrierContract(PathDependentContract):
+
+    def __init__(self, und: Stock, dtype: PutCallFwd, longshort: LongShort, strk: float, exp: float,
+                 barrier: float, updown: UpDown, inout: InOut) -> None:
```

Comment:
I would create a Barrier class for all barrier related property and functionality. Fine as it is for now, we can refactor later.

Reply:
Let us discuss.